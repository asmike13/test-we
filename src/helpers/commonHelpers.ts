import { IBook } from "../types/book"
import { ICharacter } from "../types/character"

export const getBookId = (url: string) => url.split('books/')[1]

export const getCharacterId = (url: string) => url.split('characters/')[1]
export const getCharacterIds = (urls: string[]) => urls
	.map((url) => getCharacterId(url))
	// remove undefined ids
	.filter((url) => url)

export const getCharacterName = (character?: ICharacter) => character?.name || character?.aliases[0] || ''

export const getBookFromBookList = (bookList: IBook[], id: string) => bookList.find((book) => getBookId(book.url) === id)
export const getCharacterFromCharacterList = (characterList: ICharacter[], id?: string) => characterList.find((character) => getCharacterId(character.url) === id)
