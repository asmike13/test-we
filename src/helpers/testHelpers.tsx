/* eslint-disable prefer-const */

import { Provider } from "react-redux";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { store } from "src/store";


export const getRender = (Component: () => JSX.Element, initRoute: string[] = ['/'], path?: string) => {
	return (
		<Provider store={store}>
			<MemoryRouter initialEntries={initRoute}>
				<Routes>
					<Route path="/" element={<Component />} />
					{path && <Route path={path} element={<Component />} />}
					{!path && [...initRoute].map((r, index) => (
						<Route key={index} path={r} element={<Component />} />
					))}
				</Routes>
			</MemoryRouter>
		</Provider>
	)
}

export const waitForData = (store: any, value: any) => new Promise((res) => {
	const loop = (s: any, v: any) => {
		if (JSON.stringify(store) === JSON.stringify(value)) {
			res(true)
		} else {
			setTimeout(() => {
				loop(store, value)
			}, 300);
		}
	}

	loop(store, value)
})

export const waitForElement = (
	container: HTMLElement,
	element: string,
	removed?: boolean,
	valueToEqual?: string,
	length?: number,
) => new Promise((resolve, reject) => {
	let loopTimeout: any;
	let timeout: any;
	let elementFound: any;

	const getTextContent = (el: any) => el?.textContent || el?.value;
	let isFoundAndEqual: any;

	const loop = () => {
		elementFound = valueToEqual
			? container.querySelectorAll(element)
			: container.querySelector(element);

		const isFound = elementFound && !removed && !valueToEqual;
		const isRemoved = !elementFound && removed;
		const isFoundAndLengthEqual = isFound && length && elementFound?.length === length;

		isFoundAndEqual = valueToEqual
			&& Object.values(elementFound).find((el: any) => getTextContent(el) === valueToEqual);

		if (isRemoved || isFound || isFoundAndEqual || isFoundAndLengthEqual) {
			clearTimeout(timeout);
			resolve(isFoundAndEqual || elementFound);
		} else {
			loopTimeout = setTimeout(() => {
				loop();
			}, 100);
		}
	};

	loop();

	timeout = setTimeout(() => {
		clearTimeout(loopTimeout);
		reject(Error(`${element} ${valueToEqual ? `expect ${valueToEqual} get ${getTextContent(Object.values(elementFound)[0])}` : 'not found'}`));
	}, 3000);
});

export const waitForElementToBeRemoved = (
	container: HTMLElement,
	element: string,
) => waitForElement(
	container,
	element,
	true,
);

export const waitForElementToEqual = (
	container: HTMLElement,
	element: string,
	valueToEqual: string,
) => waitForElement(
	container,
	element,
	undefined,
	valueToEqual,
);

export const waitForElementLengthToEqual = (
	container: HTMLElement,
	element: string,
	length: number,
) => waitForElement(
	container,
	element,
	undefined,
	undefined,
	length,
);
