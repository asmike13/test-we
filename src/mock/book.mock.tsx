import { IBook } from "src/types/book";

export const bookMock = {
	authors: ['string'],
	characters: ['https://anapioficeandfire.com/api/characters/1'],
	country: 'string',
	isbn: 'string',
	mediaType: 'string',
	name: 'string',
	numberOfPages: 1,
	povCharacters: ['https://anapioficeandfire.com/api/characters/1'],
	publisher: 'string',
	released: '1996-08-01T00:00:00',
	url: 'https://anapioficeandfire.com/api/books/1',
} as IBook

export const booksMock = [
	bookMock,
	{
		...bookMock,
		url: 'https://anapioficeandfire.com/api/books/2',
	},
]