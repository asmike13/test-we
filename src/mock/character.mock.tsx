import { ICharacter } from "src/types/character";

export const characterMock = {
	id: '1',
	url: 'https://anapioficeandfire.com/api/characters/1',
	name: 'string',
	gender: 'string',
	culture: 'string',
	born: 'string',
	died: 'string',
	titles: ['string'],
	aliases: ['string'],
	father: 'string',
	mother: 'string',
	spouse: 'string',
	allegiances: ['string'],
	books: ['https://anapioficeandfire.com/api/books/1'],
	povBooks: ['https://anapioficeandfire.com/api/books/1'],
	tvSeries: ['string'],
	playedBy: ['string'],
} as ICharacter

export const charactersMock = [
	characterMock,
	{
		...characterMock,
		url: 'https://anapioficeandfire.com/api/characters/2',
	}
]