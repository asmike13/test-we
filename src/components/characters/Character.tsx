import { List, Paper } from "@mui/material"
import NavBar from "../common/NavBar"
import ListItem from "../common/ListItem"
import { ICharacter } from "src/types/character"
import { getCharacterName } from "src/helpers/commonHelpers"
import Loader from "src/components/common/Loader"

interface ICharacterProps {
	characterList: ICharacter[]
	currentCharacter?: ICharacter
	characterId?: string
	status?: string
	getCharacterFromUrl: (url: string) => JSX.Element
	getBooksLink: (books: string[]) => JSX.Element[]
}

const Character = ({
	currentCharacter,
	characterId,
	status,
	getCharacterFromUrl,
	getBooksLink,
}: ICharacterProps) => {
	return (
		<>
			{status === 'pending' && <Loader />}
			{status === 'rejected' && <div className="section-title">Character not found</div>}
			{currentCharacter && currentCharacter.id === characterId &&
				<>
					<NavBar title={getCharacterName(currentCharacter)} />
					<span className="section-title">Infos :</span>
					<Paper elevation={3} sx={{ position: 'relative' }}>
						<List sx={{ marginTop: '15px', display: 'flex', flexWrap: 'wrap' }} className="character-infos-list">
							<ListItem text={`Gender: ${currentCharacter.gender}`} />
							<ListItem text={`Culture: ${currentCharacter.culture}`} />
							<ListItem text={`Born: ${currentCharacter.born}`} />
							<ListItem text={`Died: ${currentCharacter.died}`} />
							<ListItem text={`Titles: ${currentCharacter.titles}`} />
							<ListItem text={`Aliases: ${currentCharacter.aliases}`} />
							<ListItem><>Father: {getCharacterFromUrl(currentCharacter.father)}</></ListItem>
							<ListItem><>Mother: {getCharacterFromUrl(currentCharacter.mother)}</></ListItem>
							<ListItem><>Spouse: {getCharacterFromUrl(currentCharacter.spouse)}</></ListItem>
							<ListItem><>Books: {getBooksLink(currentCharacter.books)}</></ListItem>
							<ListItem><>PovBooks: {getBooksLink(currentCharacter.povBooks)}</></ListItem>
							<ListItem text={`TvSeries: ${currentCharacter.tvSeries}`} />
							<ListItem text={`PlayedBy: ${currentCharacter.playedBy}`} />
						</List>
					</Paper>
				</>
			}
		</>
	)
}

export default Character
