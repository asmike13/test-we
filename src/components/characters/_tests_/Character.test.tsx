import renderer from 'react-test-renderer';
import { getRender } from 'src/helpers/testHelpers';
import { characterMock } from 'src/mock/character.mock';
import Character from '../Character';

const props = {
	characterList: [characterMock],
	getCharacterFromUrl: jest.fn(),
	getBooksLink: jest.fn(),
}

const allProps = {
	...props,
	currentCharacter: characterMock,
	characterId: '1',
	status: 'fulfilled'
}

it('should render as snapshot with status pending', async () => {
	const snapshot = renderer.create(getRender(() => <Character status='pending' {...props} />))

	expect(snapshot).toMatchSnapshot()
})

it('should render as snapshot with status rejected', async () => {
	const snapshot = renderer.create(getRender(() => <Character status='rejected' {...props} />))

	expect(snapshot).toMatchSnapshot()
})

it('should render as snapshot with status fulfilled', async () => {
	const snapshot = renderer.create(getRender(() => <Character  {...allProps} />))

	expect(snapshot).toMatchSnapshot()
})
