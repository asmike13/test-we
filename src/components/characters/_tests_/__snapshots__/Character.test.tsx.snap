// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as snapshot with status fulfilled 1`] = `
Array [
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="HomeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
      </a>
      <a
        className="arrow-back"
        href="/"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
      </a>
      <h1
        className="MuiTypography-root MuiTypography-h1 navbar-title css-sqoje6-MuiTypography-root"
      >
        string
      </h1>
    </div>
  </header>,
  <span
    className="section-title"
  >
    Infos :
  </span>,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-obvova-MuiPaper-root"
  >
    <ul
      className="MuiList-root MuiList-padding character-infos-list css-tfgwdi-MuiList-root"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        Gender: string
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        Culture: string
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        Born: string
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        Died: string
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        Titles: string
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        Aliases: string
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        Father: 
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        Mother: 
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        Spouse: 
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        Books: 
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        PovBooks: 
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        TvSeries: string
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        PlayedBy: string
      </li>
    </ul>
  </div>,
]
`;

exports[`should render as snapshot with status pending 1`] = `
Array [
  <div
    className="overlay"
  />,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 loader css-qh7kl5-MuiPaper-root"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="HourglassBottomIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m18 22-.01-6L14 12l3.99-4.01L18 2H6v6l4 4-4 3.99V22h12zM8 7.5V4h8v3.5l-4 4-4-4z"
      />
    </svg>
  </div>,
]
`;

exports[`should render as snapshot with status rejected 1`] = `
<div>
  Book not found
</div>
`;
